### PR TITLE
Fix sparse initializer column sparsity

### PR DESCRIPTION
**Pull request by ishtihoss**
## Summary
- make nn.init.sparse zero the documented fraction of entries in each column
- align the zero-mask orientation with PyTorch's sparse initializer
- add a deterministic column-count regression test

## Tests
- python -m unittest test_init

### CODE REVIEW

**Review by zcbenz — COMMENTED** [review]
What is the equivalent API in pytorch that I can verify with?

**Comment by ishtihoss** [comment]
The equivalent PyTorch API is `torch.nn.init.sparse_`:

```python
import math
import torch
import torch.nn as nn

rows, cols = 5, 7
sparsity = 0.25

w = torch.empty(rows, cols)
nn.init.sparse_(w, sparsity=sparsity, std=1.0)

print((w == 0).sum(dim=0))
print("expected:", math.ceil(sparsity * rows))
```

PyTorch applies sparsity per column: for a `(rows, cols)` tensor it computes `ceil(sparsity * rows)` and zeros that many row entries independently in each column.

The MLX equivalent for the same parameters is:

```python
init_fn = nn.init.sparse(sparsity, mean=0.0, std=1.0)
w = init_fn(mx.zeros((rows, cols)))
```

One API difference: PyTorch's `sparse_` is in-place, hardcodes `mean=0` for the normal draw, and defaults to `std=0.01`; MLX returns a new array and exposes `mean` (default `0.0`, `std=1.0`).

**Comment by ishtihoss** [comment]
I also ran a broader copy/paste check against `torch.nn.init.sparse_`. The semantic property being checked is: for a 2D `(rows, cols)` tensor, each column has exactly `ceil(sparsity * rows)` zeroed entries.

PyTorch reference check:

```bash
python - <<'PY'
import math
import torch

print(f"torch {torch.__version__}")
shapes = [(1, 5), (2, 2), (3, 2), (5, 7), (8, 4), (11, 3)]
sparsities = [0.0, 0.1, 0.25, 0.5, 0.75, 1.0]

for rows, cols in shapes:
    for sparsity in sparsities:
        torch.manual_seed(10_000 + rows * 100 + cols)
        x = torch.empty(rows, cols)
        torch.nn.init.sparse_(x, sparsity=sparsity, std=1.0)
        expected = math.ceil(sparsity * rows)
        col_counts = (x == 0).sum(dim=0).tolist()
        assert col_counts == [expected] * cols, (
            (rows, cols), sparsity, expected, col_counts, x
        )
print(f"column-count cases passed: {len(shapes) * len(sparsities)}")

for shape in [(3,), (2, 3, 4)]:
    try:
        torch.nn.init.sparse_(torch.empty(shape), sparsity=0.5, std=1.0)
    except ValueError as exc:
        assert "2 dimensions" in str(exc)
    else:
        raise AssertionError(f"torch accepted invalid shape {shape}")
print("invalid-rank cases passed: 2")

rows, cols, sparsity = 5, 7, 0.25
torch.manual_seed(123)
x = torch.empty(rows, cols)
torch.nn.init.sparse_(x, sparsity=sparsity, std=1.0)
print("example shape=(5,7) sparsity=0.25 torch column zero counts:", (x == 0).sum(dim=0).tolist())
print("expected per column:", math.ceil(sparsity * rows))
PY
```

Matching MLX check on this branch:

```bash
python - <<'PY'
import math
import mlx.core as mx
import mlx.nn as nn

print("mlx", mx.__version__)
shapes = [(1, 5), (2, 2), (3, 2), (5, 7), (8, 4), (11, 3)]
sparsities = [0.0, 0.1, 0.25, 0.5, 0.75, 1.0]

for rows, cols in shapes:
    for sparsity in sparsities:
        mx.random.seed(10_000 + rows * 100 + cols)
        result = nn.init.sparse(sparsity=sparsity, mean=1.0, std=0.0)(
            mx.zeros((rows, cols))
        )
        expected = math.ceil(sparsity * rows)
        col_counts = mx.sum(result == 0, axis=0).tolist()
        assert col_counts == [expected] * cols, (
            (rows, cols), sparsity, expected, col_counts, result
        )
print(f"column-count cases passed: {len(shapes) * len(sparsities)}")

for shape in [(3,), (2, 3, 4)]:
    try:
        nn.init.sparse(sparsity=0.5)(mx.zeros(shape))
    except ValueError as exc:
        assert "2 dimensions" in str(exc)
    else:
        raise AssertionError(f"MLX accepted invalid shape {shape}")
print("invalid-rank cases passed: 2")

rows, cols, sparsity = 5, 7, 0.25
mx.random.seed(123)
result = nn.init.sparse(sparsity=sparsity, mean=1.0, std=0.0)(mx.zeros((rows, cols)))
print("example shape=(5,7) sparsity=0.25 MLX column zero counts:", mx.sum(result == 0, axis=0).tolist())
print("expected per column:", math.ceil(sparsity * rows))
PY
```

The output I got locally was:

```text
torch 2.12.1
column-count cases passed: 36
invalid-rank cases passed: 2
example shape=(5,7) sparsity=0.25 torch column zero counts: [2, 2, 2, 2, 2, 2, 2]
expected per column: 2

mlx 0.31.2
column-count cases passed: 36
invalid-rank cases passed: 2
example shape=(5,7) sparsity=0.25 MLX column zero counts: [2, 2, 2, 2, 2, 2, 2]
expected per column: 2
```

I used `mean=1.0, std=0.0` in the MLX check to make the intentionally zeroed entries unambiguous. PyTorch's API does not expose `mean`, so the PyTorch reference check uses deterministic seeds with `std=1.0`.

**Review by zcbenz — APPROVED** [review]
Nice fix, thanks!

**Comment by ishtihoss** [comment]
Follow-up on the failing macOS check from the previous run:

- Failed run: https://github.com/ml-explore/mlx/actions/runs/27927134303
- Failing check: `macOS (14.0)`
- Failing case: `python/tests/test_init.py::TestInit::test_sparse`, specifically `shape=(4, 3), dtype=mlx.core.float16`

The failure was introduced by the `put_along_axis` rewrite in `0bd618e6`. The sparse initializer was selecting the correct per-column row indices, but the macOS/Metal `float16` scatter-style update path did not reliably write the expected zeros for this small case. Linux/Windows jobs passed, so this appears specific to the macOS backend path used by `put_along_axis` here.

I pushed `0905e6b3` (`Avoid scatter update in sparse initializer`) to avoid that path. The initializer still selects `ceil(sparsity * rows)` random entries independently per column, but now builds a boolean rank mask and applies zeros with `mx.where` instead of scattering zeros into the sampled matrix.

Local verification on the PR worktree:

```bash
PYTHONPATH=python:python/tests /Users/porkr/projects/contributor/.venv/bin/python -m unittest python.tests.test_init.TestInit.test_sparse -v
DEVICE=cpu PYTHONPATH=python:python/tests /Users/porkr/projects/contributor/.venv/bin/python -m unittest python.tests.test_init.TestInit.test_sparse -v
DEVICE=gpu METAL_DEVICE_WRAPPER_TYPE=1 METAL_DEBUG_ERROR_MODE=0 PYTHONPATH=python:python/tests /Users/porkr/projects/contributor/.venv/bin/python -m unittest python.tests.test_init.TestInit.test_sparse -v
PYTHONPATH=python:python/tests /Users/porkr/projects/contributor/.venv/bin/python -m unittest python.tests.test_init -v
/Users/porkr/projects/contributor/.venv/bin/black --check python/mlx/nn/init.py python/tests/test_init.py
/Users/porkr/projects/contributor/.venv/bin/isort --check-only --profile=black python/mlx/nn/init.py python/tests/test_init.py
```

I also stress-checked the exact `float16` `(4, 3)` case for 10k iterations on local CPU and GPU, plus a broader CPU/GPU sweep over shapes, sparsities, and `float32`/`float16`; all passed.

A new GitHub Actions run was created for the pushed commit, but it is currently `action_required` with no jobs created yet: https://github.com/ml-explore/mlx/actions/runs/27959068848

File: python/mlx/nn/init.py [hunk 1]
```diff
@@ -384,15 +384,14 @@ def initializer(a: mx.array) -> mx.array:
...
             raise ValueError("Only tensors with 2 dimensions are supported")
 
-        rows, cols = a.shape
-        num_zeros = int(math.ceil(sparsity * cols))
+        rows = a.shape[0]
+        num_zeros = int(math.ceil(sparsity * rows))
 
-        order = mx.argsort(mx.random.uniform(shape=a.shape), axis=1)
+        order = mx.argsort(mx.random.uniform(shape=a.shape), axis=0)
+        mask = mx.argsort(order, axis=0) < num_zeros
```
Comment (zcbenz) [inline python/mlx/nn/init.py:391]: I'm not sure whether this is correct, and even if it is this is not easy to understand. Why not update `a` like the old code do?

**Review by zcbenz — CHANGES_REQUESTED** [review]

**Comment by zcbenz** [comment]
I'm closing this PR per https://github.com/ml-explore/mlx/pull/3752#issuecomment-4783471036.
